frontend-wasm/src/code_translator/mod.rs

```diff
@@ -0,0 +1,1035 @@
+            let real_dest_block = {
+                let i = state.control_stack.len() - 1 - depth;
+                let frame = &mut state.control_stack[i];
+                frame.set_branched_to_exit();
+                frame.br_destination()
+            };
+            let destination_args = state.peekn_mut(return_count);
+            builder.ins().br(real_dest_block, destination_args, span);
+        }
+        state.popn(return_count);
+    }
+    state.reachable = false;
+    Ok(())
+}
+
+/// Return the total Miden VM memory size (2^32 addresses * word (4 felts) bytes) in 64KB pages
+const fn mem_total_pages() -> i32 {
+    let bytes_fit_in_felt = 4; // although more than 32 bits can fit into a felt, use 32 bits to be safe
+    let felts_in_word = 4;
+    let total_addresses = u32::MAX as i64;
+    (total_addresses * (felts_in_word * bytes_fit_in_felt) / (64 * 1024)) as i32
+}
+
+fn translate_load(
+    ptr_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    state.push1(builder.ins().load(addr, span));
+}
+
+fn translate_load_sext(
+    ptr_ty: Type,
+    sext_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    let val = builder.ins().load(addr, span);
+    let sext_val = builder.ins().sext(val, sext_ty, span);
+    state.push1(sext_val);
+}
+
+fn translate_load_zext(
+    ptr_ty: Type,
+    zext_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let addr_int = state.pop1();
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    let val = builder.ins().load(addr, span);
+    let sext_val = builder.ins().zext(val, zext_ty, span);
+    state.push1(sext_val);
+}
+
+fn translate_store(
+    ptr_ty: Type,
+    memarg: &MemArg,
+    state: &mut FuncTranslationState,
+    builder: &mut FunctionBuilderExt<'_>,
+    span: SourceSpan,
+) {
+    let (addr_int, val) = state.pop2();
+    let val_ty = builder.func().dfg.value_type(val);
+    let arg = if ptr_ty != *val_ty {
+        builder.ins().trunc(val, ptr_ty.clone(), span)
+    } else {
+        val
+    };
+    let addr = prepare_addr(addr_int, &ptr_ty, memarg, builder, span);
+    builder.ins().store(addr, arg, span);
+}
+
+fn prepare_addr(
+    addr_int: Value,
+    ptr_ty: &Type,
+    memarg: &MemArg,
+    builder: &mut FunctionBuilderExt,
+    span: SourceSpan,
+) -> Value {
+    let full_addr_int = if memarg.offset != 0 {
+        let offset = builder.ins().i32(memarg.offset as i32, span);
+        builder.ins().add(addr_int, offset, span)
```

Comment:
```suggestion
        builder.ins().add_imm(addr_int, Immediate::U32(memarg.offset as u32), span)
```

You'll need to use `u32` for the type of a pointer-as-integer, and likewise for offsets (handling negative offsets as subtracting, rather than adding), since signed 32-bit arithmetic is not implemented yet.

In general I would use `getelementptr` for this kind of thing, however it currently only works on pointers to aggregates, I will probably extend it to support indexing off primitive types as well though to allow offsetting pointers "raw". We could then allow signed offsets, and rely on the backend to work out the most efficient lowering for the effective address

Reply:
Done. I left `inttoptr` for now.